rclrs/src/node/client.rs

```diff
@@ -0,0 +1,257 @@
+use crate::node::client::oneshot::Canceled;
+use futures::channel::oneshot;
+use std::borrow::Borrow;
+use std::boxed::Box;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::sync::atomic::{AtomicI64, Ordering};
+use std::sync::Arc;
+
+use crate::error::{RclReturnCode, ToResult};
+use crate::MessageCow;
+use crate::Node;
+use crate::{rcl_bindings::*, RclrsError};
+
+use parking_lot::{Mutex, MutexGuard};
+use rosidl_runtime_rs::Message;
+
+// SAFETY: The functions accessing this type, including drop(), shouldn't care about the thread
+// they are running in. Therefore, this type can be safely sent to another thread.
+unsafe impl Send for rcl_client_t {}
+
+/// Internal struct used by clients.
+pub struct ClientHandle {
+    rcl_client_mtx: Mutex<rcl_client_t>,
+    rcl_node_mtx: Arc<Mutex<rcl_node_t>>,
+}
+
+impl ClientHandle {
+    pub(crate) fn lock(&self) -> MutexGuard<rcl_client_t> {
+        self.rcl_client_mtx.lock()
+    }
+}
+
+impl Drop for ClientHandle {
+    fn drop(&mut self) {
+        let handle = self.rcl_client_mtx.get_mut();
+        let rcl_node_mtx = &mut *self.rcl_node_mtx.lock();
+        unsafe {
+            rcl_client_fini(handle as *mut _, rcl_node_mtx as *mut _);
+        }
+    }
+}
+
+impl From<Canceled> for RclrsError {
+    fn from(_: Canceled) -> Self {
+        RclrsError::RclError {
+            code: RclReturnCode::Error,
+            msg: None,
+        }
+    }
+}
+
+/// Trait to be implemented by concrete Client structs
+/// See [`Client<T>`] for an example
+pub trait ClientBase: Send + Sync {
+    /// Internal function to get a reference to the `rcl` handle.
+    fn handle(&self) -> &ClientHandle;
+    /// Tries to take a new response and run the callback or future with it.
+    fn execute(&self) -> Result<(), RclrsError>;
+}
+
+type RequestValue<Response> = Box<dyn FnOnce(&Response) + 'static + Send>;
+
+/// Main class responsible for publishing data to ROS topics
+pub struct Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    pub(crate) handle: Arc<ClientHandle>,
+    requests: Mutex<HashMap<i64, RequestValue<T::Response>>>,
+    futures: Arc<Mutex<HashMap<i64, oneshot::Sender<T::Response>>>>,
+    sequence_number: AtomicI64,
+}
+
+impl<T> Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    /// Creates a new client.
+    pub fn new(node: &Node, topic: &str) -> Result<Self, RclrsError>
+    where
+        T: rosidl_runtime_rs::Service,
+    {
+        let mut rcl_client = unsafe { rcl_get_zero_initialized_client() };
+        let type_support = <T as rosidl_runtime_rs::Service>::get_type_support()
+            as *const rosidl_service_type_support_t;
+        let topic_c_string = CString::new(topic).unwrap();
+        let rcl_node = &mut *node.rcl_node_mtx.lock();
```

Comment:
Does the mutex need to be locked for the entire rest of the function? Or could we restrict the lock to a smaller block?

Would this be overkill? My thought is that the node is potentially having quite a few things trying to initialize at the same time, and there might be several locks being attempted. Shortening the length of time a lock is active might help to keep the initialization times down.

Reply:
Done in a49d6caf54b7844d0f31bab71e150bfb23330aa7